blogs/treewidth.md

```diff
@@ -0,0 +1,385 @@
+
+Using the language of tensor network, we can rewrite the above theorem as follows: the bottleneck of time complexity of the contraction of a tensor network is $O(e^{O(tw(L(G)))})$, where $L(G)$ is the line graph of the hypergraph of the tensor network. Therefore, if we can find the tree decomposition of the tensor network with minimal treewidth, we can find the optimal contraction order of the tensor network.
+
+To further make use of the theorem, we need to introduce the concepts of line graph, tree width and tree decomposition.
+
+### Line Graph
+
+A formal definition of the [**line graph**](https://en.wikipedia.org/wiki/Line_graph) is as follows:
+
+**Definition 1**. Given a graph G, its line graph $L(G)$ is a graph such that: 
+   1. each vertex of $L(G)$ represents an edge of $G$; 
+   2. two vertices of $L(G)$ are adjacent if and only if their corresponding edges share a common endpoint ("are incident") in $G$. 
+
+That is, it is the intersection graph of the edges of G, representing each edge by the set of its two endpoints.
+
+For a tensor network, we can construct a hypergraph $G$ whose vertices are the tensors and whose hyperedges are the indices. Then the line graph $L(G)$ of the hypergraph $G$ is the graph whose vertices are the indices and whose edges are the tensors, which is a simple graph, as shown in the following figure:
+
+![Fig.1](/assets/treewidth_figs/linegraph.png)
+
+Since we are considering a tensor network, dimension of the indices have to be considered. 
+Therefore, for each vertex of the line graph $L(G)$, we define its weight as $\log_2(d)$, where $d$ is the dimension of the index. 
+
+In this way, any possible tensor can be represented as a subgraph induced by a set of vertices in $L(G)$, and its log2 size is given by the sum of weights of the boundary vertices of the subgraph. For example, the tensor $C$ in the figure above corresponding to the subgraph induced by the vertices $\{b, d, f\}$, and its log2 size is 
+$${dims(b) \times dims(d) \times dims(f)} = 2^{w(b) + w(d) + w(f)}$$
+where $w(*)$ for weights of vertices.
+
+### Tree Decomposition and Tree Width
+
+Intuitively, a [**tree decomposition**](https://en.wikipedia.org/wiki/Tree_decomposition) represents the vertices of a given graph $G$ as subtrees of a tree, in such a way that vertices in $G$ are adjacent only when the corresponding subtrees intersect.
+
+The tree decomposition of a graph is a tree whose nodes are subsets of the vertices of the graph, and the following conditions are satisfied:
+
+1. Each vertex of the graph is in at least one node of the tree.
+2. For each edge of the graph, there is a node of the tree containing both vertices of the edge.
+3. Bags containing the same vertex have to be connected in the tree.
+
+An example of tree decomposition is shown in the following figure:
+
+![](/assets/treewidth_figs/treedecomposition.png)
+
+where the left graph is the original graph and the right one is the tree decomposition of the graph.
```

Comment:
Is it clear how to construct such tree graphs? Eventually the goal is to construct the tree with minimal width of the line graph of the tensor network contraction graph; it is not clear to me whether the graph in the left of this example is even a valid line graph of any tensor network contraction. It might be more clear for the reader if there is a single example that you use consistently throughout the post?

Reply:
In the appendix, we used a new example that corresponds to an actual tensor network.